tests/bosonic_files/test_backend_bosoniccircuit.py

```diff
@@ -0,0 +1,406 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+Unit tests for backends.bosonicbackend.bosoniccircuit.py.
+"""
+
+import numpy as np
+import strawberryfields.backends.bosonicbackend.bosoniccircuit as circuit
+import pytest
+
+pytestmark = pytest.mark.bosonic
+
+NUM_WEIGHTS_VALS = [1, 2, 5]
+NUM_MODES_VALS = [1, 3, 4]
+NUM_NEW_WEIGHTS_VALS = [1, 2, 3]
+R_VALS = np.array([-1.0, -1 / 3, 1 / 2, 1.2])
+PHI_VALS = np.array([-1.2, -1 / 5, 0.6, 1.3])
+NBAR_VALS = np.array([0, 1, 2.5])
+NUM_SHOTS_VALS = np.array([1, 2, 10])
+R_ANC_VALS = np.array([0.1, 0.8, 1.2, 5])
+ETA_ANC_VALS = np.array([0.1, 0.4, 0.9, 1])
+
+
+class TestBosonicCircuit:
+    r"""Tests all the functions inside backends.bosonicbackend.bosoniccircuit.py"""
+
+    @pytest.mark.parametrize("num_weights", NUM_WEIGHTS_VALS)
+    @pytest.mark.parametrize("num_modes", NUM_MODES_VALS)
+    def test_reset_circuit(self, num_weights, num_modes):
+        r"""Checks that the reset method instantiates correct number of modes
+        as vacuum, each with a given number of weights."""
+        # Create a state over 3 modes, 2 weights per mode
+        example = circuit.BosonicModes()
+        example.reset(3, 2)
+
+        # Confirm number of modes and weights
+        assert example.nlen == 3
+
+        # Confirm normalization
+        assert np.isclose(sum(example.sweights()), 1)
+
+        # Confirm each mode initialized to vacuum
+        assert example.is_vacuum(tol=1e-10)
+        for i in range(3):
+            assert np.isclose(example.fidelity_vacuum([i]), 1)
+
+        # Confirm active modes
+        assert example.active == [0, 1, 2]
+        # Check mode-wise and quadrature-wise ordering
+        assert np.allclose(example.to_xp, np.array([0, 2, 4, 1, 3, 5]))
+        assert np.allclose(example.from_xp, np.array([0, 3, 1, 4, 2, 5]))
+
+        # Confirm number of weights, means and covs
+        assert example.sweights().shape == (2 ** 3,)
+        assert example.smean().shape == (2 ** 3, 6)
+        assert example.scovmat().shape == circuit.c_shape(3, 2)
+
+        # Reset with number of weights and modes and perform the same check
+        example.reset(num_modes, num_weights)
+        assert example.nlen == num_modes
+        assert np.isclose(sum(example.sweights()), 1)
+        assert example.is_vacuum(tol=1e-10)
+        for i in range(num_modes):
+            assert np.isclose(example.fidelity_vacuum([i]), 1)
+        assert example.active == [i for i in range(num_modes)]
+        assert example.smean().shape == (num_weights ** num_modes, 2 * num_modes)
+        assert example.scovmat().shape == (num_weights ** num_modes, 2 * num_modes, 2 * num_modes)
+
+    @pytest.mark.parametrize("num_weights", NUM_WEIGHTS_VALS)
+    @pytest.mark.parametrize("num_modes", NUM_MODES_VALS)
+    @pytest.mark.parametrize("num_weights_new", NUM_NEW_WEIGHTS_VALS)
+    def test_add_del_modes(self, num_weights, num_modes, num_weights_new):
```

Comment:
There are two separate methods tested in here: addition and deletion. Suggest splitting them up.

Reply:
I purposefully tested them together because they both change the list of active modes. I wanted to ensure that no conflict was created between them.

Reply:
This way, it is not ensured that deletion and addition work separately with the test. It could arise that some effects that are caused by deletion are pruned by the followed addition, which could cause the test to pass falsely.

Reply:
Will split.